python/ray/serve/_private/long_poll.py

```diff
@@ -398,10 +405,14 @@ def notify_changed(
         object_key: KeyType,
         updated_object: Any,
     ):
-        self.snapshot_ids[object_key] += 1
+        try:
+            self.snapshot_ids[object_key] += 1
+        except KeyError:
+            # Initial snapshot id must be >= 0, so that the long poll client
+            # can send a negative initial snapshot id to get a fast update.
+            self.snapshot_ids[object_key] = 0
```

Comment:
This was previously a random number https://github.com/ray-project/ray/pull/45881/files#diff-f138b21f7ddcd7d61c0b2704c8b828b9bbe7eb5021531e2c7fabeb20ec322e1aL195 but I couldn't figure out why - I can of course use a random number here but this seems simpler if that's not necessary.

Reply:
@shrekris-anyscale do you know why this was a random number / is that necessary?

Reply:
We set it to a random number to handle an edge case when the controller crashes and recovers. Suppose the controller always starts with ID 0, and it's currently at ID 1. Suppose all the clients are also at ID 1. Now suppose:

1. The `ServeController` (which runs the `LongPollHost`) crashes and restarts. All snapshot IDs are reset to 0.
2. All clients– except 1 slow client– reconnect to the controller. Their snapshot IDs are still 1, so the controller propagates an update, and all the connected clients' snapshot IDs are set back to 0.
3. One of the connected clients sends an update to the `LongPollHost` using `notify_changed`. The controller bumps the snapshot ID to 1 and updates the connected clients.
4. The slow client finally connects to the controller. Its ID is also 1, so **it doesn't receive the update.**

To correctly avoid this edge case, we should cache the snapshot IDs and restore them when the controller recovers. That's pretty complex though, so instead we initialize the snapshot IDs to a random number between 0 and 1,000,000. That makes this edge case very unlikely.

Could we switch back to using a default dict with the random number generator as the factory function?

Reply:
Oh! That totally makes sense, and I will restore that behavior.

> Could we switch back to using a default dict with the random number generator as the factory function?

Let me play around with this, I might be able to make it work

Reply:
Oh I just remembered why I changed it do a plain dict - the problem is that `snapshot_ids` is a defaultdict, but `object_snapshots` isn't, so I was trying to get items from one that didn't exist in the other. Now I'm wondering if those two mappings can be combined, since they have the same keys 🤔